pkg/network/go/bininspect/pclntab.go

```diff
@@ -0,0 +1,284 @@
+	symbolFilter symbolFilter
+
+	// byteOrderParser is the binary.ByteOrder for the pclntab.
+	byteOrderParser binary.ByteOrder
+	// cachedVersion is the version of the pclntab.
+	cachedVersion version
+	// funcNameTable is the sectionAccess for the function name table.
+	funcNameTable sectionAccess
+	// funcData is the sectionAccess for the function data.
+	funcData sectionAccess
+	// funcTable is the sectionAccess for the function table.
+	funcTable sectionAccess
+	// funcTableSize is the size of the function table.
+	funcTableSize uint32
+	// ptrSize is the size of a pointer in the architecture of the binary.
+	ptrSize uint32
+	// ptrBufferSizeHelper is a buffer for reading pointers of the size ptrSize.
+	ptrBufferSizeHelper []byte
+	// funcNameHelper is a buffer for reading function names. Of the maximum size of the symbol names.
+	funcNameHelper []byte
+	// funcTableFieldSize is the size of a field in the function table.
+	funcTableFieldSize int
+	// funcTableBuffer is a buffer for reading fields in the function table.
+	funcTableBuffer []byte
+}
+
+// GetPCLNTABSymbolParser returns the matching symbols from the pclntab section.
+func GetPCLNTABSymbolParser(f *elf.File, symbolFilter symbolFilter) (map[string]*elf.Symbol, error) {
+	section := f.Section(pclntabSectionName)
+	if section == nil {
+		return nil, ErrMissingPCLNTABSection
+	}
+
+	parser := &pclntanSymbolParser{section: section, symbolFilter: symbolFilter}
+
+	if err := parser.parsePclntab(); err != nil {
+		return nil, err
+	}
+	// Late initialization, to prevent allocation if the binary is not supported.
+	_, maxSymbolsSize := symbolFilter.getMinMaxLength()
+	parser.funcNameHelper = make([]byte, maxSymbolsSize)
+	parser.funcTableFieldSize = getFuncTableFieldSize(parser.cachedVersion, int(parser.ptrSize))
+	// Allocate the buffer for reading the function table.
+	// TODO: Do we need 2*funcTableFieldSize?
+	parser.funcTableBuffer = make([]byte, 2*parser.funcTableFieldSize)
+	return parser.getSymbols()
+}
+
+// parsePclntab parses the pclntab, setting the version and verifying the header.
+// Based on parsePclnTab in https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L194
+func (p *pclntanSymbolParser) parsePclntab() error {
+	p.cachedVersion = ver11
+
+	pclntabHeader := make([]byte, 8)
+	if n, err := p.section.ReadAt(pclntabHeader, 0); err != nil || n != len(pclntabHeader) {
+		return fmt.Errorf("failed to read pclntab header: %w", err)
+	}
+	// Matching the condition https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L216-L220
+	// Check header: 4-byte magic, two zeros, pc quantum, pointer size.
+	if p.section.Size < 16 || pclntabHeader[4] != 0 || pclntabHeader[5] != 0 ||
+		(pclntabHeader[6] != 1 && pclntabHeader[6] != 2 && pclntabHeader[6] != 4) || // pc quantum
+		(pclntabHeader[7] != 4 && pclntabHeader[7] != 8) { // pointer size
+		// TODO: add explicit error message
```

Comment:
Unclear what needs to be explict? The read value of pclntabheader should be included in the message?

Reply:
Thought about adding the exact failure, like `section size is smaller than 16`, or `pointer size is invalid`, etc.